lib/cn/kcompact.c

```diff
@@ -175,176 +79,212 @@ kcompact(struct cn_compaction_work *w, struct kvset_builder *bldr)
+                &vdata, &vlen, &complen)) {
+            bool should_emit = false;
+
+            /* Assertion logic:
+             *   if (dbg_nvals_this_key)
+             *       assert(dbg_prev_seq > seq);
+             */
+            if (HSE_UNLIKELY(dbg_nvals_this_key && dbg_prev_seq <= seq)) {
+                assert(0);
+                seqno_errcnt++;
             }
 
-            if (w->cw_drop_tombs && (vtype == vtype_tomb || vtype == vtype_ptomb))
-                continue; /* skip value */
-        }
+            dbg_nvals_this_key++;
+            dbg_prev_seq = seq;
 
-        if (vtype == vtype_ptomb)
-            should_emit = !emitted_seq_pt || seq < emitted_seq_pt;
-        else
-            should_emit = !emitted_seq || seq < emitted_seq;
+            if (seq <= w->cw_horizon) {
+                horizon = false;
+                if (pt_set && seq < pt_seq)
+                    continue; /* skip value */
 
-        should_emit = should_emit || !emitted_val;
+                if (vtype == vtype_ptomb) {
+                    pt_set = true;
+                    pt_kobj = curr->kobj;
+                    assert(key_obj_len(&curr->kobj) == w->cw_pfx_len);
+                    pt_seq = seq;
+                }
 
-        /* Compare seq to emitted_seq to ensure when a key has values
-         * in two kvsets with the same sequence number, that only the
-         * value from the first kvset is emitted.
-         */
-        if (should_emit) {
-            switch (vtype) {
+                if (w->cw_drop_tombs && (vtype == vtype_tomb || vtype == vtype_ptomb))
+                    continue; /* skip value */
+            }
+
+            if (vtype == vtype_ptomb)
+                should_emit = !emitted_seq_pt || seq < emitted_seq_pt;
+            else
+                should_emit = !emitted_seq || seq < emitted_seq;
+
+            should_emit = should_emit || !emitted_val;
+
+            /* Compare seq to emitted_seq to ensure when a key has values
+             * in two kvsets with the same sequence number, that only the
+             * value from the first kvset is emitted.
+             */
+            if (should_emit) {
+                switch (vtype) {
                 case vtype_val:
                 case vtype_cval:
                     err = kvset_builder_add_vref(
-                        bldr, seq, vbidx + w->cw_vbmap.vbm_map[curr.src],
-                        vboff, vlen, complen);
+                        bldr, seq, vbidx + w->cw_vbmap.vbm_map[idx], vboff, vlen, complen);
                     break;
                 case vtype_zval:
                 case vtype_ival:
-                    err = kvset_builder_add_val(bldr, &curr.kobj, vdata, vlen, seq, 0);
+                    err = kvset_builder_add_val(bldr, &curr->kobj, vdata, vlen, seq, 0);
                     break;
                 default:
                     err = kvset_builder_add_nonval(bldr, seq, vtype);
                     break;
+                }
```

Comment:
Curious that these "vtypes" are not capitalized.  And as Alex always point out, don't use default case with enumerated types (I realize this predates your changes).

Reply:
Yeah I would like to change this at some point as well.